### ISSUE TITLE
Text not appearing in the images exported to .pdf from .svg using asciidoctor-pdf.

### ISSUE THREAD

**Issue by saurabhsingh-pvips** [mogest/prawn-svg#113]
This is the rendered image I get in the pdf

![image](https://user-images.githubusercontent.com/57518239/70896497-d4378880-2016-11ea-9667-fcef52a47f29.png)

VERSUS 

The image in the html file of the same .adoc .
![image](https://user-images.githubusercontent.com/57518239/70896570-ee716680-2016-11ea-933e-93c94bc31760.png)

The raw SVG file can be found [here](https://github.com/sifive/block-pio-sifive/blob/d125018d0b4f24ca5ceb1887b74121eccebf5076/docs/scribble/components/pio/pio-diagram.svg)

asciidoctor-pdf used: 

```
Asciidoctor PDF 1.5.0.rc.1.dev using Asciidoctor 2.0.10 [https://asciidoctor.org]
Runtime Environment (ruby 2.6.5p114 (2019-10-01 revision 67812) [x86_64-linux]) (lc:UTF-8 fs:UTF-8 in:UTF-8 ex:UTF-8)
```

**Comment by mogest** [comment]
Had a quick look.  It seems that this SVG puts its `<defs>` at the bottom of document, and uses the elements defined in the `<defs>` before they're actually defined.

That doesn't appear to be invalid SVG, but it's not recommended—see the SVG spec:

> It is recommended that, where possible, referenced elements be defined prior to the elements that use them, in document order. Collecting all referenced elements inside of a single ‘defs’ element near the top of the file can make the markup easier to read and understand.

The fix would have to be scanning the document for `<defs>` before we start parsing, which is going to slow down every render.  I'll do some performance tests to see whether this is going to have a noticeable effect.

If you can make whatever is generating the SVG put the `<defs>` above the uses of those defs, you'll have fixed the problem you're experiencing at least...

**Comment by mogest** [comment]
Hi @saurabhsingh-pvips, turned out to be not too much work.  It seems to render your sample SVG perfectly.  I probably won't do a release only for this fix but please do pull down from master and try it out.

**Comment by saurabhsingh-pvips** [comment]
Alright. I did try manually editing the svg file but couldn't succeed completely. I am currently facing issue in the steps to generate .adoc file and it might take a while. Once generated, would try again with the master version and let know if it's fixed. Thanks!

**Comment by saurabhsingh-pvips** [comment]
How do I pull prawn-svg directly from master into the asciidoctor-pdf?

**Comment by mojavelinux** [comment]
@saurabhsingh-pvips Create the following Gemfile in your project:

```rb
source 'https://rubygems.org'

gem 'asciidoctor-pdf'
gem 'prawn-svg', github: 'mogest/prawn-svg'
```

Run Bundler as follows:

```
$ bundle --path=.bundle/gems
```

Run asciidoctor-pdf via Bundler:

```
bundle exec asciidoctor-pdf doc.adoc
```

**Comment by saurabhsingh-pvips** [comment]
This falls back to the previous error of `allow-uri-read is not enabled`. I had previously fixed this by pulling master of asciidoctor-pdf and setting up from there.

**Comment by saurabhsingh-pvips** [comment]
Using this version of asciidoctor-pdf falls back to previous error of `allow-uri-read not enabled`. I had previously fixed this by setting up asciidoctor-pdf installed from master of asciidoctor-pdf.

**Comment by mojavelinux** [comment]
Then also use the development version of asciidoctor-pdf:

```rb
gem 'asciidoctor-pdf', github: 'asciidoctor/asciidoctor-pdf'
```

**Comment by saurabhsingh-pvips** [comment]
Stuck at error:
`asciidoctor: WARNING: could not embed image:`
```
Missing end tag for 'g' (got "defs")
Line: 1
Position: 9595
Last 80 unconsumed characters:
```
Complete error message here
[asciidoctor.txt](https://github.com/mogest/prawn-svg/files/3997536/asciidoctor.txt)

**Comment by mojavelinux** [comment]
Looks to me like the XML is invalid. Have you checked it with an XML syntax checker?

**Comment by saurabhsingh-pvips** [comment]
Checked with two online validators and there's no error!

**Comment by mojavelinux** [comment]
I tried it and it works fine. I downloaded the SVG in the description. Then I used:

```
image::index.svg[pdfwidth=100%]
```

And I ran that with:

```
bundle exec asciidoctor-pdf test.adoc
```

I see the image exactly as it should appear.

**Comment by saurabhsingh-pvips** [comment]
I am not sure where I should use this: `image::index.svg[pdfwidth=100%]`

**Comment by mojavelinux** [comment]
@saurabhsingh-pvips it seems like your error is coming from somewhere else.

@mogest I can confirm that the change you made allows the text to appear.

**Comment by mojavelinux** [comment]
@saurabhsingh-pvips you are giving bits and pieces of your scenario, so I don't really know what to suggest. But it sounds like you have having difficulty with Asciidoctor PDF at this point and not prawn-svg. I would recommend either asking on gitter.im/asciidoctor/asciidoctor or filing an issue at asciidoctor-pdf for help. I don't think we need to concern mogest with it.

**Comment by saurabhsingh-pvips** [comment]
I guess my scenarios is pretty straight forward. I've an adoc file with svg image embedded. I need to convert it to pdf. We've validated the svg so I can narrow the issue down to something not being right with asciidoctor-pdf. 

Another thing to be noted. I did fresh installation of asciidoctor-pdf again and the error message shown are the same, even without specifying it to use the development version. Are you sure you checked out the latest commit of asciidoctor before running this or are you using some old version which might be working?

I'll file an issue at gitter.

**Comment by mojavelinux** [comment]
> Are you sure you checked out the latest commit of asciidoctor before running this or are you using some old version which might be working?

I'm absolutely positive. You can be certain of that. I work with these tools every day so I'm keenly aware of what versions I'm using.

> I've an adoc file with svg image embedded. 

The problem is, you are not giving us that example. You attached an SVG to the description of this issue and that SVG works perfectly. So it's likely the fact that your scenario is different. And that is best suited for an issue in Asciidoctor PDF, not here.

**Comment by saurabhsingh-pvips** [comment]
I get it. This might be something related with asciidoctor-pdf. I'll file a ticket there.

**Comment by saurabhsingh-pvips** [comment]
Some how it is fixed. I restarted the terminal and somehow it worked.

**Comment by mojavelinux** [comment]
Yeah! :tada:
